File: solutions/zig/proverb/1/proverb.zig
const std = @import("std");
const fmt = std.fmt;
const mem = std.mem;

const verse = "For want of a {s} the {s} was lost.\n";
const ending = "And all for the want of a {s}.\n";

pub fn recite(allocator: mem.Allocator, words: []const []const u8) (fmt.AllocPrintError || mem.Allocator.Error)![][]u8 {
    var arr = std.ArrayList([]u8).init(allocator);

    if (words.len > 0) {
        if (words.len > 1) {
            var it = mem.window([]const u8, words[0..], 2, 1);
            while (it.next()) |w|
                try arr.append(try fmt.allocPrint(allocator, verse, .{ w[0], w[1] }));
        }
        try arr.append(try fmt.allocPrint(allocator, ending, .{words[0]}));
    }

    return arr.toOwnedSlice();
}

File: solutions/lua/proverb/1/proverb.lua
local verse = "For want of a %s the %s was lost.\n"
local ending = "And all for the want of a %s.\n"

return {
  recite = function(strings)
    local res = ""
    for i = 2, #strings do
      res = res .. verse:format(strings[i - 1], strings[i])
    end
    if #strings > 0 then res = res .. ending:format(strings[1]) end
    return res
  end
}
